### PR TITLE
CT: Allow configuration of acceptable EKUs

### LINKED ISSUE #606: CT: Allow configuration of acceptable EKUs
Currently `x509.ExtKeyUsageAny` is used in `x509.VerifyOptions` which means any EKU will be accepted (including i.e. S/MIME or code signing certs that cannot be used for server auth). For production CA use we would most likely want to restrict submitted certificates to those that can be used for server authentication so we'd want to just set `x509.ExtKeyUsageServerAuth`.

I think there is some value in being able to fine tune this to some degree (beyond just flipping a all/server-auth only switch) but how to allow configuring this can get a little complicated.

### PR DESCRIPTION
**Pull request by rolandshoemaker**
Fixes #606

### CODE REVIEW

File: examples/ct/handlers.go [hunk 1]
```diff
@@ -159,6 +159,8 @@ type LogContext struct {
 	trustedRoots *PEMCertPool
 	// rejectExpired indicates whether certificate validity period should be used during chain verification
 	rejectExpired bool
+	// extKeyUsages contains the list of EKUs to use during chain verification
+	extKeyUsages []x509.ExtKeyUsage
```
Comment (daviddrysdale) [inline examples/ct/handlers.go]: Sorry to be a pain, but I think we've reached the point where this stuff would be better in a separate sub-structure.  If we added something like:
```go
type CertValidationOpts struct {
 	// trustedRoots is a pool of certificates that defines the roots the CT log will accept
￼  	trustedRoots *PEMCertPool
￼  	// rejectExpired indicates whether certificate validity period should be used during chain verification
￼  	rejectExpired bool
￼ 	// extKeyUsages contains the list of EKUs to use during chain verification
￼ 	extKeyUsages []x509.ExtKeyUsage
}
```
we could use that everywhere (as a field in `LogContext` and as parameter to `ValidateChain`)

Comment (rolandshoemaker) [inline examples/ct/handlers.go]: This definitely makes sense, will do.

File: examples/ct/cert_checker_test.go [hunk 2]
```diff
@@ -94,8 +94,7 @@ func TestCertCheckerInvalidChainAccepted(t *testing.T) {
 		t.Fatal("failed to load fake root")
 	}
 
-	_, err := ValidateChain(jsonChain, *trustedRoots, false)
-
+	_, err := ValidateChain(jsonChain, CertValidationOpts{trustedRoots, false, []x509.ExtKeyUsage{x509.ExtKeyUsageAny}})
```
Comment (daviddrysdale) [inline examples/ct/cert_checker_test.go]: (nit) Maybe have a `testValidationOpts` at the top of the file, with field names broken out, and use it throughout?

File: examples/ct/instance.go [hunk 3]
```diff
@@ -82,6 +84,21 @@ func LogConfigFromFile(filename string) ([]LogConfig, error) {
...
+	"IPSECUser":                  x509.ExtKeyUsageIPSECUser,
+	"TimeStamping":               x509.ExtKeyUsageTimeStamping,
+	"OCSPSigning":                x509.ExtKeyUsageOCSPSigning,
+	"MicrosoftServerGatedCrypto": x509.ExtKeyUsageMicrosoftServerGatedCrypto,
+	"NetscapeServerGatedCrypto":  x509.ExtKeyUsageNetscapeServerGatedCrypto,
+}
+
 // SetUpInstance sets up a log instance that uses the specified client to communicate
 // with the Trillian RPC back end.
 func (cfg LogConfig) SetUpInstance(client trillian.TrillianLogClient, deadline time.Duration) (*PathHandlers, error) {
```
Comment (daviddrysdale) [inline examples/ct/instance.go]: Could you add `TODO(daviddrysdale): add unit tests` at the top of this function while you're in here?

**Comment by daviddrysdale** [comment]
To save another round-trip, I subsumed the first 2 commits from this into pull #616 and added some test changes.

Probably best to run `./scripts/presubmit --no-generate` before pushing though :-)

**Pull request closed without merging**